core/meta/src/TClass.cxx

```diff
@@ -5459,35 +5471,39 @@ TClass *TClass::LoadClassCustom(const char *requestedname, Bool_t silent)
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-/// Try to load the classInfo (it may require parsing the header file
-/// and/or loading data from the clang pcm).
+/// Try to load the ClassInfo if available. This function may require parsing
+/// the header file and/or loading data from the clang pcm. If further calls to
+/// this function cannot affect the value of fClassInfo, fCanLoadClassInfo is set
+/// to false.
 
 void TClass::LoadClassInfo() const
 {
-   R__LOCKGUARD(gInterpreterMutex);
+   bool autoParse = !gInterpreter->IsAutoParsingSuspended();
 
-   // If another thread executed LoadClassInfo at about the same time
-   // as this thread return early since the work was done.
-   if (!fCanLoadClassInfo) return;
+   {
+      R__LOCKGUARD(gInterpreterMutex);
```

Comment:
Both `AutoParse()` and `SetClassInfo()` below lock `gInterpreterMutex` themselves. Is there a point to taking the lock here for them to be coherent (i.e., avoid two waits for the lock), or should we simply remove the lock and let it be taken by the interpreter? It seems better to still take the lock, and actually extend it to the remaining of the function to ensure that no work is duplicated by other threads (that is, extend it until we set fCanLoadClassInfo to false, to trigger the early return in other threads).